include/drivers/qspi.h

```diff
@@ -0,0 +1,660 @@
+/*
+ * Copyright (c) 2019, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/**
+ * @file
+ * @brief Public APIs for QSPI drivers and applications
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_QSPI_H_
+#define ZEPHYR_INCLUDE_DRIVERS_QSPI_H_
+
+/**
+ * @defgroup qspi_interface QSPI Interface
+ * @ingroup io_interfaces
+ * @brief QSPI Interface
+ *
+ * The QSPI API provides support for the standard flash
+ * memory devices
+ * @{
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*
+ * The following #defines are used to configure the QSPI controller.
+ */
+
+/**
+ * @brief No of data lines that are used for the transfer.
+ * Some flash memories supports one, two or four lines.
+ */
+#define QSPI_DATA_LINES_SINGLE                  0x00
+#define QSPI_DATA_LINES_DOUBLE                  0x01
+#define QSPI_DATA_LINES_QUAD                    0x02
+
+/**
+ * @brief QSPI mode of the transmission
+ * For detailed description see mode in qspi_config
+ */
+#define QSPI_MODE_0                             0x00
+#define QSPI_MODE_1                             0x01
+#define QSPI_MODE_2                             0x02
+#define QSPI_MODE_3                             0x03
+
+/**
+ * @brief QSPI Address configuration
+ * Length of the address field in bits.
+ * Typical flash chips support 24bit address mode
+ */
+#define QSPI_ADDRESS_MODE_8BIT                  0x00
+#define QSPI_ADDRESS_MODE_16BIT                 0x01
+#define QSPI_ADDRESS_MODE_24BIT                 0x02
+#define QSPI_ADDRESS_MODE_32BIT                 0x03
+
+/**
+ * @brief QSPI buffer structure
+ * Structure used both for TX and RX purposes.
+ *
+ * @param buf is a valid pointer to a data buffer.
+ * Can not be NULL.
+ * @param len is the length of the data to be handled.
+ * If no data to transmit/receive - pass 0.
+ */
+struct qspi_buf {
+	u8_t *buf;
+	size_t len;
+};
+
+
+/**
+ * @brief QSPI command structure
+ * Structure used for custom command usage.
+ *
+ * @param op_code is a command value (i.e 0x9F - get Jedec ID)
+ * @param tx_buf structure used for TX purposes. Can be NULL if not used.
+ * @param rx_buf structure used for RX purposes. Can be NULL if not used.
+ */
+struct qspi_cmd {
```

Comment:
This struct should contain more information as more parameters might to be set per command:
-cmd: opcode, buswidth
-addr: nbytes, buswidth
-...

Reply:
I can agree up to a point. In my opinion parameters like buswidth, address and others should be set by `configure` function. It is not without reason - when you want to change bus parameters, in general you have to uninit and then reinitialise perihperial. Proposed API contains function for that functionality - it is `configure`. 
In your proposition function `qspi_send_cmd` does the same as  `configure`. In my opinion each function has to do unique things.

Reply:
I can agree up to a point. In my opinion parameters like buswidth, address and others should be set by `configure` function. It is not without reason - when you want to change bus parameters, in general you have to uninit and then reinitialise perihperial. Proposed API contains function for that functionality - it is `configure`.
In your proposition function `qspi_send_cmd `does the same as `configure`. In my opinion each function has to do unique things.

Reply:
> In your proposition function qspi_send_cmd does the same as configure. In my opinion each function has to do unique things.

No, I just have a different split of parameters. Some parameters you consider static for a given memory might actually change within a single transaction.

Reply:
Could you give an example? I have already work with some qspi memories and I have not found any situation to change `address  `size field (of course in case of single memory on the bus).
The only case when `data_lines` has to be reconfigured is when you are enabling Quad mode, but it is done only once.

Reply:
I have the following example for a read on spi-nand memory (but which would require the exact same "spi memory" api), you need the following operations:
CMD PAGE_READ, nbytes 3,
CMD GET_FEATURE, nybytes 1,
CMD PAGE_READ_FROM_CACHE_OP, nbytes 2, 

Also note that for instance GET_FEATURE expects response on bus width of 1 line, even if memory has been configured in quad mode. Which justify the need to have the buswidth info at each command.

Reply:
Thank you for the example! 
Indeed it shows situation when single data transfer is required. But it still do not touches the problem of changing the `address` field.
What do you think about this solution:
1. `qspi_write`/`qspi_read `commands operate with defined number of `data_lines` (i.e four)
2. `qspi_send_cmd` always operate with single data line.
That solution enables using fast four line transfer for read/write commands and still provide functionality for single line communication (i.e read status register).